src/VisualStudio/Core/Def/Implementation/TableDataSource/VisualStudioDiagnosticListTable.cs

```diff
@@ -130,5 +150,98 @@ private void OnErrorListPropertyChanged(object sender, PropertyChangedEventArgs
                 AddTableSourceIfNecessary(this.Workspace.CurrentSolution);
             }
         }
+
+        private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs e)
+        {
+            Contract.ThrowIfFalse(_errorList is IErrorList2);
+
+            switch (e.Kind)
+            {
+                case WorkspaceChangeKind.SolutionAdded:
+                case WorkspaceChangeKind.SolutionChanged:
+                case WorkspaceChangeKind.SolutionCleared:
+                case WorkspaceChangeKind.SolutionReloaded:
+                case WorkspaceChangeKind.SolutionRemoved:
+                case WorkspaceChangeKind.ProjectAdded:
+                case WorkspaceChangeKind.ProjectChanged:
+                case WorkspaceChangeKind.ProjectRemoved:
+                    // Set error list toggle state based on current analysis state for all languages for projects in current solution.
+                    var fullAnalysisState = _optionService.GetOption(RuntimeOptions.FullSolutionAnalysis);
+                    if (fullAnalysisState)
+                    {
+                        var langauges = e.NewSolution.Projects.Select(p => p.Language).Distinct();
+                        foreach (var language in langauges)
+                        {
+                            if (!ServiceFeatureOnOffOptions.IsClosedFileDiagnosticsEnabled(_optionService, language))
+                            {
+                                fullAnalysisState = false;
+                                break;
+                            }
+                        }
+                    }
+
+                    ((IErrorList2)_errorList).AnalysisToggleState = fullAnalysisState;
+                    return;
+            }
+        }
+
+        private void OnErrorListFullSolutionAnalysisToggled(object sender, AnalysisToggleStateChangedEventArgs e)
+        {
+            var newOptions = _optionService.GetOptions()
```

Comment:
You can just write _workspace.Options = workspace.Options.WithChangedOption(....). I've been trying to move people off from the internal APIs to public ones.